patches/tModLoader/Terraria/ModLoader/ModGore.cs

```diff
@@ -1,117 +1,39 @@
 using Microsoft.Xna.Framework;
-using System;
-using System.Collections.Generic;
-using Terraria.GameContent;
-using Terraria.ID;
-using Terraria.ModLoader.Core;
 
 namespace Terraria.ModLoader
 {
 	/// <summary>
-	/// This class allows you to customize the behavior of a custom gore. Create a new instance of this and pass it as a parameter to Mod.AddGore to customize the gore you are adding. If you are autoloading gore, then give it the same name as the gore texture.
+	/// This class allows you to customize the behavior of a custom gore.
 	/// </summary>
-	public class ModGore
+	public class ModGore : ModTexturedType
 	{
-		private static int nextGore = GoreID.Count;
-		public static int GoreCount => nextGore;
-		internal static readonly IDictionary<string, int> gores = new Dictionary<string, int>();
-		internal static readonly IDictionary<int, ModGore> modGores = new Dictionary<int, ModGore>();
-
 		/// <summary> Allows you to copy the Update behavior of a different type of gore. This defaults to 0, which means no behavior is copied. </summary>
 		public int updateType = -1;
 
-		internal static int ReserveGoreID() {
-			int reserveID = nextGore;
-			nextGore++;
-			return reserveID;
-		}
-
-		/// <summary>
-		/// Gets the type of the custom gore corresponding to the given texture. Returns 0 if the texture does not represent a gore.
-		/// </summary>
-		public static int GetGoreSlot(string texture) {
-			if (gores.ContainsKey(texture)) {
-				return gores[texture];
-			}
-			else {
-				return 0;
-			}
-		}
-
-		//in Terraria.GameContent.ChildSafety make SafeGore internal and not readonly
-		internal static void ResizeAndFillArrays() {
-
-			//Textures
-			Array.Resize(ref TextureAssets.Gore, nextGore);
-
-			//Sets
-			LoaderUtils.ResetStaticMembers(typeof(GoreID), true);
-			Array.Resize(ref ChildSafety.SafeGore, nextGore);
-
-			for (int k = GoreID.Count; k < nextGore; k++) {
-				GoreID.Sets.DisappearSpeed[k] = 1;
-				GoreID.Sets.DisappearSpeedAlpha[k] = 1;
-			}
-
-			foreach (string texture in gores.Keys) {
-				TextureAssets.Gore[gores[texture]] = ModContent.GetTexture(texture);
-			}
-		}
-
-		internal static void Unload() {
-			gores.Clear();
-			modGores.Clear();
-			nextGore = GoreID.Count;
-		}
+		internal string nameOverride;
+		internal string textureOverride;
 
-		//in Terraria.Gore add modGore property (internal set)
-		//in Terraria.Gore.NewGore after resetting properties call ModGore.SetupGore(Main.gore[num]);
-		internal static void SetupGore(Gore gore) {
-			if (modGores.ContainsKey(gore.type)) {
-				gore.modGore = modGores[gore.type];
-				gore.modGore.OnSpawn(gore);
-			}
-			else {
-				gore.modGore = null;
-			}
-		}
+		public int Type { get; internal set; }
 
-		internal static void SetupUpdateType(Gore gore) {
-			if (gore.modGore != null && gore.modGore.updateType > 0) {
-				gore.realType = gore.type;
-				gore.type = gore.modGore.updateType;
-			}
-		}
+		public override string Name => nameOverride ?? base.Name;
```

Comment:
can we do a `ModGoreSimple` subclass instead?

Reply:
GoreLoader.AddGoreFromTexture<T> is currently usable with custom ModGores too.

Reply:
good point